File: 3-oauth2-google-login/1-java/src/main/java/com/example/oauthdemo/service/AuthService.java
package com.example.oauthdemo.service;

import com.example.oauthdemo.entity.User;
import com.example.oauthdemo.repository.UserRepository;
import com.example.oauthdemo.security.JwtService;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
public class AuthService {
    private final UserRepository userRepository;
    private final JwtService jwtService;

    public AuthService(UserRepository userRepository, JwtService jwtService) {
        this.userRepository = userRepository;
        this.jwtService = jwtService;
    }

    public Map<String, Object> findOrCreateGoogleUser(String googleId, String email, String firstName, String lastName, String picture) {
        Optional<User> userOpt = userRepository.findByGoogleId(googleId);
        if (userOpt.isEmpty()) {
            userOpt = userRepository.findByEmail(email);
        }

        User user;
        boolean isNewUser;
        if (userOpt.isEmpty()) {
            user = new User();
            user.setEmail(email);
            user.setGoogleId(googleId);
            user.setFirstName(firstName);
            user.setLastName(lastName);
            user.setPicture(picture);
            user.setPassword(null);
            user = userRepository.save(user);
            isNewUser = true;
        } else {
            user = userOpt.get();
            user.setGoogleId(googleId);
            if (firstName != null) user.setFirstName(firstName);
            if (lastName != null) user.setLastName(lastName);
            if (picture != null) user.setPicture(picture);
            user = userRepository.save(user);
            isNewUser = false;
        }

        Map<String, Object> result = new HashMap<>();
        result.put("user", user);
        result.put("isNewUser", isNewUser);
        return result;
    }

    public Map<String, Object> completeGoogleLogin(User user, boolean isNewUser) {
        String accessToken = jwtService.generateToken(user.getId());

        String name = Stream.of(user.getFirstName(), user.getLastName())
                .filter(s -> s != null && !s.isEmpty())
                .collect(Collectors.joining(" "));

        Map<String, Object> userMap = new HashMap<>();
        userMap.put("id", user.getId());
        userMap.put("email", user.getEmail());
        userMap.put("name", name);
        userMap.put("googleSub", user.getGoogleId());

        Map<String, Object> response = new HashMap<>();
        response.put("message", "Đăng nhập Google thành công!");
        response.put("access_token", accessToken);
        response.put("isNewUser", isNewUser);
        response.put("user", userMap);

        return response;
    }
}

File: 3-oauth2-google-login/0-typescript/src/modules/auth/auth.service.ts
/**
 * Business logic service for Auth.
 */
import {
    Injectable,
} from "@nestjs/common"
import {
    JwtService,
} from "@nestjs/jwt"
import {
    InjectRepository,
} from "@nestjs/typeorm"
import {
    Repository,
} from "typeorm"
import {
    UserEntity,
} from "../../entities"
import type {
    GoogleProfilePayload,
} from "./google-profile"

/**
 * Result tuple — isNewUser=true on INSERT, false on lookup.
 */
export interface FindOrCreateResult {
    user: UserEntity
    isNewUser: boolean
}

/**
 * Links Google profiles to DB rows and signs internal JWTs.
 */
@Injectable()
export class AuthService {
    constructor(
    @InjectRepository(UserEntity)
    private readonly usersRepo: Repository<UserEntity>,
    private readonly jwtService: JwtService,
    ) {}

    /**
     * Silent registration / linking: if googleId/email is new → INSERT (isNewUser=true);
     * Upserts local user linked to Google identity and reports whether a new row was created.
     *
     * @param payload — normalized Google payload
     * @returns Cặp persisted user + new-row flag
     */
    async findOrCreateGoogleUser(payload: GoogleProfilePayload): Promise<FindOrCreateResult> {
        let user =
      (await this.usersRepo.findOne({
          where: {
              googleId: payload.googleId,
          },
      })) ??
      (await this.usersRepo.findOne({
          where: {
              email: payload.email,
          },
      }))

        if (!user) {
            user = this.usersRepo.create({
                email: payload.email,
                googleId: payload.googleId,
                firstName: payload.firstName ?? null,
                lastName: payload.lastName ?? null,
                picture: payload.picture ?? null,
                password: null,
            })
            await this.usersRepo.save(user)
            return {
                user,
                isNewUser: true,
            }
        }

        user.googleId = payload.googleId
        user.firstName = payload.firstName ?? user.firstName
        user.lastName = payload.lastName ?? user.lastName
        user.picture = payload.picture ?? user.picture
        await this.usersRepo.save(user)
        return {
            user,
            isNewUser: false,
        }
    }

    /**
     * Backwards-compatible wrapper returning only the user entity.
     */
    async findOrCreateFromGoogle(payload: GoogleProfilePayload): Promise<UserEntity> {
        const { user } = await this.findOrCreateGoogleUser(payload)
        return user
    }

    /**
     * Issues JWT with sub claim and returns isNewUser for client onboarding split.
     *
     * @param user — persisted user entity
     * @param isNewUser — first-time login flag
     */
    async completeGoogleLogin(user: UserEntity, isNewUser: boolean): Promise<{
        message: string
        access_token: string
        isNewUser: boolean
        user: { id: number; email: string; name: string; googleSub: string | null }
    }> {
        const access_token = await this.jwtService.signAsync({
            sub: user.id,
        })
        return {
            message: "Đăng nhập Google thành công!",
            access_token,
            isNewUser,
            user: {
                id: user.id,
                email: user.email,
                name: [user.firstName,
                    user.lastName,
                ].filter(Boolean).join(" "),
                googleSub: user.googleId,
            },
        }
    }
}
